### ISSUE TITLE
 Rsyslog omkafka queue problem

### ISSUE THREAD

**Issue by mhyllander** [rsyslog/rsyslog#1052]
The title of this issue is the same as #796, but since it never got resolved I decided to open a new one.

I am facing the same problem as the original author. I have not been able to get the DA queue that I have configured for the omkafka action to start saving logs to disk when the kafka broker is down. Neither does it save the queue to disk when rsyslog is stopped.

I've been running rsyslog in debug mode, and it looks to me like the logs are being de-queued and passed to librdkafka. The queue is therefore empty and there are no logs to save.

Digging around the librdkafka docs, I see this:

> rd_kafka_produce() is a non-blocking API, it will en-queue the message on an internal queue and return immediately. If the number of queued messages would exceed the "queue.buffering.max.messages" configuration property then rd_kafka_produce() returns -1 and sets errno to ENOBUFS, thus providing a backpressure mechanism.

In other words, librdkafka has an internal queue which basically "swallows" a large amount of logs, `queue.buffering.max.messages` is by default 100000! This means that if rsyslog is stopped while the broker is down, you can lose up to 100000 logs.

Indeed, if I set `confParam=["queue.buffering.max.messages=1"]` in the omkafka action, the action queue will start to fill up and eventually start saving logs to spool files. The save on shutdown also works. Unfortunately syslog starts filling up with messages like this:

```
2016-05-26T23:59:27.455477+02:00 hive-vag-vag-ipserv-ip1-0 rsyslogd-2359: action 'action 0' resumed (module 'omkafka') [v8.19.0.master try http://www.rsyslog.com/e/2359 ]
2016-05-26T23:59:27.455630+02:00 hive-vag-vag-ipserv-ip1-0 rsyslogd-2424: omkafka: Failed to produce to topic 'logs' partition -1: Local: Queue full  [v8.19.0.master try http://www.rsyslog.com/e/2424 ]
2016-05-26T23:59:27.455638+02:00 hive-vag-vag-ipserv-ip1-0 rsyslogd-2359: action 'action 0' resumed (module 'omkafka') [v8.19.0.master try http://www.rsyslog.com/e/2359 ]
2016-05-26T23:59:27.455662+02:00 hive-vag-vag-ipserv-ip1-0 rsyslogd-2424: omkafka: Failed to produce to topic 'logs' partition -1: Local: Queue full  [v8.19.0.master try http://www.rsyslog.com/e/2424 ]
2016-05-26T23:59:27.455671+02:00 hive-vag-vag-ipserv-ip1-0 rsyslogd-2359: action 'action 0' resumed (module 'omkafka') [v8.19.0.master try http://www.rsyslog.com/e/2359 ]
2016-05-26T23:59:27.455673+02:00 hive-vag-vag-ipserv-ip1-0 rsyslogd-2424: omkafka: Failed to produce to topic 'logs' partition -1: Local: Queue full  [v8.19.0.master try http://www.rsyslog.com/e/2424 ]
2016-05-26T23:59:27.455677+02:00 hive-vag-vag-ipserv-ip1-0 rsyslogd-2359: action 'action 0' resumed (module 'omkafka') [v8.19.0.master try http://www.rsyslog.com/e/2359 ]
2016-05-26T23:59:27.455679+02:00 hive-vag-vag-ipserv-ip1-0 rsyslogd-2424: omkafka: Failed to produce to topic 'logs' partition -1: Local: Queue full  [v8.19.0.master try http://www.rsyslog.com/e/2424 ]
2016-05-26T23:59:27.455683+02:00 hive-vag-vag-ipserv-ip1-0 rsyslogd-2359: action 'action 0' resumed (module 'omkafka') [v8.19.0.master try http://www.rsyslog.com/e/2359 ]
2016-05-26T23:59:27.455685+02:00 hive-vag-vag-ipserv-ip1-0 rsyslogd-2424: omkafka: Failed to produce to topic 'logs' partition -1: Local: Queue full  [v8.19.0.master try http://www.rsyslog.com/e/2424 ]
2016-05-26T23:59:27.455689+02:00 hive-vag-vag-ipserv-ip1-0 rsyslogd-2359: action 'action 0' resumed (module 'omkafka') [v8.19.0.master try http://www.rsyslog.com/e/2359 ]
2016-05-26T23:59:27.455691+02:00 hive-vag-vag-ipserv-ip1-0 rsyslogd-2424: omkafka: Failed to produce to topic 'logs' partition -1: Local: Queue full  [v8.19.0.master try http://www.rsyslog.com/e/2424 ]
```

I suppose the same thing would happen when the queue is full, regardless of the `queue.buffering.max.messages` value. However, with queue size = 1 performance is _extremely_ bad (librdkafka of course uses its internal queue for batching requests to the broker), so this is no solution.

I also read that librdkafka is asynchronous in nature (https://github.com/edenhill/librdkafka/issues/39#issuecomment-31338158, a related issue), and that:

> It is thus not safe to discard a message at produce() time because a partition is not available: the application might have started to produce message's before the initial broker connection has been set up, or before all relevant brokers have been connected to - ....

I venture to guess that the problem in omkafka is that logs are de-queued immediately after being sent to librdkafka. Ideally this should be delayed until a callback indicates that the logs have been successfully transmitted to the broker, but I don't know if that is possible.

On my Ubuntu 14.04 I have librdkafka 0.8.3, but I see in the librdkafka github project that more configuration parameters have been added in 0.9.1. Maybe some of them could help with reliability, e.g. `max.in.flight.requests.per.connection` or `message.timeout.ms`?

**Comment by linuxmalaysia** [comment]
Im using kafka, this issue is a concern to us.

**Comment by xrl** [comment]
I'm worried about this as well. Our rsyslog error logs are filling up with these issues.

**Comment by radu-gheorghe** [comment]
@mhyllander, maybe using a callback to check whether messages actually got delivered will work? Like shown here? https://github.com/edenhill/librdkafka/wiki/Sync-producer

**Comment by mhyllander** [comment]
Yes, I was also thinking that the delivery report callback could be used. How I would like it to work is for omkafka to enqueue messages to librdkafka for delivery, but *not* dequeue them from the rsyslog queue until the delivery callback indicates that the messages have been delivered. That way, undelivered messages would still be in the rsyslog queue and could be saved to disk if rsyslog is shutdown.

**Comment by samriang** [comment]
Any chances this issue to be fixed in near future?
Trying to build reliable logs delivery with omkafka.

**Comment by samriang** [comment]
Any news about this issue?

**Comment by samriang** [comment]
Knock-knock. Any ideas about DA for omkafka?

**Comment by rgerhards** [comment]
We need someone who has time and interest in enhancing omkafka to provide a status back to rsyslog (whatever it takes).

**Comment by missnebun** [comment]
+1

**Comment by oldmantaiter** [comment]
Going to wrap my head around it and see if there is anything I can do. It has been a while so I'll need to get up to speed with changes since 7.4.7 (last time I was poking around)

**Comment by samriang** [comment]
@oldmantaiter It sounds very encouraging :)

**Comment by pemontto** [comment]
Big +1 on this, we are also using omhiredis which was just recently improved to properly use the rsyslog queue and ensure message delivery and retries.

Confirmed I lost lots of logs when bouncing rsyslog while Kafka was down. As suggested I've minimised this impact by adding "queue.buffering.max.messages=10000", but much smaller than that and the throughput drops substantially.

**Comment by davidelang** [comment]
what do you have the retries set for? IIRC the default is to not retry failed deliveries.

**Comment by pemontto** [comment]
```
        action.resumeRetryCount="-1"
        action.resumeInterval="5"
        action.reportSuspension="on"
        action.reportSuspensionContinuation="on"
```

**Comment by rgerhards** [comment]
@davidelang The problem is not rsyslog queue. The problem is that kafka accepts and acks messages even though they might still get lost. So we would need to modify omkafka to wait for a callback from kafka that all has actually been comitted (as suggested by @radu-gheorghe above). However, I guess this, if done naively, will also lead to performance issues. So I think we need the help from someone who actually is into kafka develoopment.

**Comment by oldmantaiter** [comment]
I haven't actually had any time yet (newborn at home) to look into this - maybe someone else might want to carry this torch since I may not have a lot of free time in the next little while.

We may also want to engage the author of librdkafka and ask for his opinion on what might be the best course of action, perhaps something can be added to the library that would make this easier...

**Comment by samriang** [comment]
Hope to see the solution someday :)

**Comment by frostynova** [comment]
Hope +1

**Comment by blakambr** [comment]
Sending in my +1

**Comment by samriang** [comment]
Can anybody try to set
`action.resumeretrycount="-1"`
for omkafka action and look into main DA queue when high watermark will be passed?

**Comment by davidelang** [comment]
As I am understanding the problem here, the issue is that omkafka is telling the 
rsyslog that the message was successfully delivered, even if the far end tells 
omkafka that it failed. As such, setting retries = -1 won't solve the problem.

Someone who knows kafka will need to look at the contributed omkafka module 
source and figure out what it's doing wrong so that it can tell rsyslog that the 
message was not successfully delivered.

David Lang

P.S. as I understand things, kafka provides application level acknowledgements, 
so omkafka needs to model itself after omrelp which also has app-level acks 
rather than omfwd which does not.

**Comment by samriang** [comment]
@davidelang Do you know how the omkafka handling messages if the destination host is unreachable at all?

**Comment by davidelang** [comment]
On Mon, 9 Jan 2017, samriang wrote:

> @davidelang Do you know how the omkafka handling messages if the destination host is unreachable at all?

no, I have not looked into that.

David Lang

**Comment by davidelang** [comment]
I took a quick look at the omkafka queue and it looks like it's just calling a libkafka function to add the message to the libkafka queue.

It looks like this module really needs a in-depth rework, possibly even eliminating the use of libkafka and having the module deliver messages to kafka over the network so that it can track delivery directly.

librelp with it's confirmation delivery is the place to start.

**Comment by rgerhards** [comment]
just FYI: I am currently talking with a potential sponsor for an in-depth overhaul of omkafka, and am positive this will happen. If so, we will get a real great upgrade for the module. Just bear with me a couple of more days, then I should know more.

**Comment by leoluk** [comment]
Glad to hear. In its current form, it's basically unusable except if you don't care about delivery guarantees at all (that limitation should probably be noted in the docs to prevent people from relying on it when they shouldn't...).

I recommend pushing directly to Kafka using Elastic's filebeat without having yet another message queue in between. The log file on 

Kafka is perfect for that use case and handles thousands of producers just fine. In my tests, I easily got up to 200k msgs/s per broker without tuning with ~1k producers. You'll have to set up authentication to prevent producers from consuming the queue.

If you need to process syslog input, use rsyslog and log to a file, then ship the file to Kafka using filebeat

**Comment by xrl** [comment]
@rgerhards any news on the kafka refresh?

**Comment by rgerhards** [comment]
Actually we got the sponsorship late last week. So expect  mayor work the coming weeks...

**Comment by feiyuewuyue** [comment]
I want to know if the bug can resolve at version 8.28.0

**Comment by modeyang** [comment]
+1

**Comment by rgerhards** [comment]
as you can see, the fix was applied to 8.28.0, and this is why the tracker is closed. So this is solved.
